src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DoublyLinkedList.cs

```diff
@@ -0,0 +1,186 @@
+
+            _last._next = other._first;
+            other._first._prev = _last;
+            _last = other._last;
+            _size += other._size;
+
+            other._size = -1;
+        }
+
+        /// <summary>Append all the elements from all the other lists at the end of this list (O(others.Length) operation, the other lists must be discarded)</summary>
+        public void Append(DoublyLinkedList<T>[] others)
+        {
+            for (int i = 0; i < others.Length; ++i)
+            {
+                Append(others[i]);
+            }
+        }
+
+        /// <summary>Insert the given element at the end of this list (O(1) operation)</summary>
+        public void InsertAtEnd(T elem)
+        {
+            Debug.Assert(CheckValidity());
+
+            if (_last is null)
+            {
+                //this list is empty
+                _first = new(elem);
+                _last = _first;
+                _size = 1;
+                return;
+            }
+
+            _last._next = new(elem, _last, null);
+            _last = _last._next;
+            _size += 1;
+        }
+
+        /// <summary>Insert the given element at the start of this list (O(1) operation)</summary>
+        public void InsertAtStart(T elem)
+        {
+            Debug.Assert(CheckValidity());
+
+            if (_first is null)
+            {
+                //this list is empty
+                _first = new(elem);
+                _last = _first;
+                _size = 1;
+                return;
+            }
+
+            _first._prev = new(elem, null, _first);
+            _first = _first._prev;
+            _size += 1;
+        }
+
+        /// <summary>Enumerate all elements in the list</summary>
+        /// <param name="inreverse">if true then the enumeration happens in reverse order</param>
+        public IEnumerable<T> Enumerate(bool inreverse = false) => inreverse ? EnumerateBackwards() : EnumerateForwards();
+
+        private IEnumerable<T> EnumerateForwards()
+        {
+            Debug.Assert(CheckValidity());
+
+            Node? current = _first;
+            while (current is not null)
+            {
+                yield return current._elem;
+                current = current._next;
+            }
+        }
+
+        private IEnumerable<T> EnumerateBackwards()
+        {
+            Debug.Assert(CheckValidity());
+
+            Node? current = _last;
+            while (current is not null)
+            {
+                yield return current._elem;
+                current = current._prev;
+            }
+        }
+
+        private sealed class Node
+        {
+            public Node? _next;
+            public Node? _prev;
+            public readonly T _elem;
+
+            public Node(T elem) { _elem = elem; }
+
+            public Node(T elem, Node? prev, Node? next) { _elem = elem; _prev = prev; _next = next; }
+        }
+
+        public override string ToString()
```

Comment:
can this ToSTring be debug only?

Reply:
yes

Reply:
Indeed, both ToString()'s are used for debugging purposes only, forgot to mark them as such.

Reply:
I could do a quick follow-up fix PR, unless someone else takes care of this inside another PR, as this is a trivial change that perhaps does not warrant a separate PR.

Reply:
Well, I'll remember to do it in my upcoming PR (regarding upper bound measuring of NFA states).